agentrs/src/ws/client.rs

```diff
@@ -0,0 +1,190 @@
+use crate::{
+    conf,
+    tasks::*,
+    ws::{
+        rdp_message_processor::MessageProcessor,
+        types::{ChannelMap, ProxyMap, SessionMap, WsWriter},
+    },
+};
+use anyhow::Context;
+use async_trait::async_trait;
+use axum::http::HeaderValue;
+use std::collections::HashMap;
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::sync::RwLock;
+use tracing::{debug, error, info};
+use tungstenite::{client::IntoClientRequest, handshake::client::Request};
+
+use futures::{SinkExt, StreamExt};
+use tokio::sync::Mutex;
+use tokio_tungstenite::{connect_async, tungstenite::protocol::Message};
+
+#[derive(Clone)]
+pub struct WebSocket {
+    pub config_manager: conf::ConfigHandleManager,
+    pub request: Request,
+    pub reconnect_interval: Duration,
+}
+
+#[async_trait]
+impl Task for WebSocket {
+    type Output = anyhow::Result<()>;
+
+    const NAME: &'static str = "agent";
+
+    async fn run(self, mut shutdown_signal: ShutdownSignal) -> Self::Output {
+        tokio::select! {
+            result = self.run_with_reconnect() => result,
+            _ = shutdown_signal.wait() => Ok(()),
+        }
+    }
+
+    fn get_name(&self) -> &'static str {
+        Self::NAME
+    }
+}
+
+impl WebSocket {
+    pub fn new() -> anyhow::Result<Self> {
+        let config_manager =
+            conf::ConfigHandleManager::init().context("Failed to init config manager")?;
+
+        let gateway_url = std::env::var("GATEWAY_URL")
+            .unwrap_or_else(|_| "ws://localhost:8009/api/ws".to_string());
+        let mut request = gateway_url.into_client_request().unwrap();
+
+        // Insert a custom header
+        let token = config_manager.conf.token.clone().unwrap();
+        let token: &'static str = Box::leak(token.into_boxed_str());
+        request
+            .headers_mut()
+            .insert("HOOP_KEY", HeaderValue::from_static(token));
```

Comment:
Maybe something like this, avoids too much allocation on the heap.
```suggestion
        let token = config_manager.conf.token.clone().unwrap();
        request
            .headers_mut()
            .insert("HOOP_KEY", HeaderValue::from_str(token.as_str()).unwrap());
```